File: arrays_dynamicarr.py
#Dynamic Array
class DynamicArray:
    def __init__(self):
        self.capacity= 2
        self.size= 0
        self.data= [None]*self.capacity
        
    def resize(self):
        self.capacity*= 2
        
        new_data= [None]* self.capacity
        
        
        for i in range(self.size):
            new_data[i]= self.data[i]
            
        self.data= new_data
        
    def push_back(self, value):
        if self.size==self.capacity:
            self.resize()
            
        self.data[self.size]= value
        self.size+= 1
        
    def pop_back(self):
        if self.size==0:
            raise IndexError("Array is empty")
        
        value= self.data[self.size-1]
        
        self.data[self.size-1]= None
        self.size-=1
        
        return value
    
    def get(self, index):
        if index<0 or index>=self.size:
            raise IndexError("Index out of bounds")
        
        return self.data[index]
    
    def set(self, index, value):
        if index<0 or index>=self.size:
            raise IndexError("Index out of bounds")
        
        self.data[index]= value
        
        
# Circular Buffer
class CircularBuffer:
    def __init__(self, capacity):
        self.capacity = capacity
        self.size=0
        self.head=0
        self.tail=0
        self.data= [None]*self.capacity
        
    def empty(self):
        return self.size==0
    
    def full(self):
        return self.size==self.capacity
    
    def push_back(self,value):
        if self.full():
            raise OverflowError("Buffer full")
        
        self.data[self.tail]= value
        self.tail= (self.tail+1)%self.capacity
        
        self.size+=1
        
    def pop_front(self):
        if self.empty():
            raise IndexError("Buffer is empty")
        
        self.head= (self.head+1)% self.capacity
        self.size-=1
        
    def front(self):
        if self.empty():
            raise IndexError("Buffer is Empty")
        
        return self.data[self.head]

File: Arrays/arrays_dynamicarr.cpp
/* 
Dynamic Array Implementation

APIs: Constructor, Destructor, Copy Constructor, Assginment Operator, push_back, pop_back(), get, [], set, size, empty
*/

template <typename T>
class DynamicArray
{
    private:
        T* data;
        int size;
        int cap;

        void resize()
        {
            int cap= cap*2;

            T* newData= new T[cap];

            for (int i=0; i<size; i++)
            {
                newData[i]= data[i];
            }

            delete[] data;
            data= newData;
        }

    public:

        DynamicArray()
        {
            cap=2;
            size=0;
            data= new T[cap];
        }

        ~DynamicArray()
        {
            delete[] data;
        }

        DynamicArray(const DynamicArray& other)
        {
            if (this==&other) return *this;
            
            delete[] data;

            size= other.size;
            cap= other.cap;
            data= new T[cap];

            for (int i=0; i<size; i++)
            {
                data[i]= other.data[i];
            }

            return *this;
        }

        void push_back(const T& val)
        {
            if (size==cap) resize();

            data[size]= value;
            size++;
        }

        void pop_back()
        {
            if (size==0) throw std::out_of_range("Empty Arr");

            size--;
        }

        T get(int index) const
        {
            if (index<0 || index>size)
            {
                throw std::out_of_range("Index out of range");
            }
            return data[index];
        }

        void set(int index, const T& val)
        {
            if (inedx<0 || index>size)
            {
                throw std::out_of_range("Index out of range");
            }

            data[index]= value;
        }

        bool empty() const
        {
            return size==0;
        }

        T& operator[](int index)
        {
            if (index<0 || index>=size)
            {
                throw std::out_of_range("Index out of bounds")
            }
            return data[index];
        }


};





/*
Circular Buffer

API: pop_front, push_back, empty, full, size, capacity, front
*/

template <typename T>
class CircularBuffer
{
    private:
        T* data;
        int cap;
        int size;
        int head;
        int tail;

    public:
        CircularBuffer(int capacity)
        {
            cap= capacity;
            size= 0;
            data= new T[cap];
            head= 0;
            tail= 0;
        }

        ~CircularBuffer()
        {
            delete[] data;
        }

        bool empty() const
        {
            return size==0;
        }

        bool full() const
        {
            return size==cap;
        }
        
        // [1], h=0, t=1 [1,2,3,4], t=0
        void push_back(const T& value)
        {
            if (full())
            {
                throw std::overflow_error("full");
            }
            data[tail]= value;
            tail= (tail+1)%cap;

            size++;
        }

        // [1,2,3,4], h=0, t=0, h=1
        void pop_front()
        {
            if (empty())
            {
                throw std::underflow_error("Buffer is empty");
            }
            head= (head+1)%cap;

            size--;
        }

        T front() const
        {
            if (empty())
            {
                throw std::underflow_error("Buffer is empty");
            }
            return data[head];
        }
};